### PR TITLE
Add has and delete methods to WeakMap

### PR DESCRIPTION
**Pull request by thoov**

### CODE REVIEW

File: packages/ember-metal/lib/weak_map.js [hunk 1]
```diff
@@ -37,3 +37,24 @@ WeakMap.prototype.set = function(obj, value) {
...
   return this;
 };
+
+/*
+ * @method has
+ * @param key {Object}
+ * @return {Boolean} if the key exists
+ */
+WeakMap.prototype.has = function(obj) {
+  return !!this.get(obj);
```
Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: what if the value being stored is `false` or `0`

Comment (thoov) [inline packages/ember-metal/lib/weak_map.js]: Ok will change

Comment (thoov) [inline packages/ember-metal/lib/weak_map.js]: @stefanpenner fixed

**Comment by thoov** [comment]
Also added check for keys that are not objects when setting

File: packages/ember-metal/lib/weak_map.js [hunk 2]
```diff
@@ -34,6 +41,39 @@ WeakMap.prototype.get = function(obj) {
  * @return {Any} stored value
  */
 WeakMap.prototype.set = function(obj, value) {
+  assert('Uncaught TypeError: Invalid value used as weak map key', typeOf(obj) === 'object');
+
+  if (typeOf(value) === 'undefined') {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: `value === undefined` is what we want here.

File: packages/ember-metal/lib/weak_map.js [hunk 3]
```diff
@@ -34,6 +41,39 @@ WeakMap.prototype.get = function(obj) {
...
+
+/*
+ * @method has
+ * @param key {Object}
+ * @return {Boolean} if the key exists
+ */
+WeakMap.prototype.has = function(obj) {
+  var map = meta(obj).readableWeak();
+
+  if (map && typeOf(map[this._id]) !== 'undefined') {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: ^

File: packages/ember-metal/lib/weak_map.js (hunk 2 above)
Comment (thoov) [inline packages/ember-metal/lib/weak_map.js]: good catch. I updated

File: packages/ember-metal/lib/weak_map.js [hunk 4]
```diff
@@ -1,7 +1,10 @@
+import { typeOf } from 'ember-runtime/utils';
+import { assert } from 'ember-metal/debug';
 import { GUID_KEY } from 'ember-metal/utils';
 import { meta } from 'ember-metal/meta';
 
 var id = 0;
+const UNDEFINED = function() {};
```
Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: lets do `function UNDEFINED() {}` that way it is named nicely in most modern debuggers.

Comment (thoov) [inline packages/ember-metal/lib/weak_map.js]: good call. updated

**Comment by stefanpenner** [comment]
thanks :)

**Pull request merged**